### ISSUE TITLE
Identifier 'global' has already been declared

### ISSUE THREAD

**Issue by egilsster** [nats-io/stan.js#104]
When running a test that imports the module, it causes a duplicate identifier error to be thrown. This seems to be the cause https://github.com/nats-io/node-nats-streaming/blob/master/lib/pb/protocol_pb.js#L10

Project to reproduce the issue https://github.com/egilsster/nats-bug

**Comment by aricart** [comment]
The protobuf compiler pollutes the namespace with `global` and other names. Looks like we may have a fix - https://github.com/protocolbuffers/protobuf/issues/4271.

**Comment by egilsster** [comment]
@aricart changing `commonjs` to `commonjs_strict` results in the `global` variable being called `proto`. I think changing that might require some extra work but cant really say, I am by no means familiar with the code here but I could take an honest stab at it but would appreciate some help with getting this issue fixed. 🙏

**Comment by aricart** [comment]
@egilsster I have been :) - still working on it, the issue is that the fix they provided for commonjs_strict changed the behaviour to generate a bunch of files one per type - however these files are not directly consumable, they need to be processed by 2 additional tools. I am having trouble with the google closure compiler to not get stumbled on the protobuf library dependencies, see the referenced issue above.

Realize that the pollution problem is still on your plate for the other protobuf? library :)

**Comment by egilsster** [comment]
Yeah I am going to do a hacky fix for that until this gets resolved – its because of my unit test runner (Jest).

**Comment by ss37** [comment]
> Yeah I am going to do a hacky fix for that until this gets resolved – its because of my unit test runner (Jest).

@egilsster What was the workaround that helped in the meantime?

**Comment by egilsster** [comment]
@egilsster since it was only viable for my unit-tests, I renamed the variable in a file which I then replaced in a pre-test step and in a post-test step I reverted it.

For my component tests, unfortunately, I had to write an isolated test and run that one with `mocha`. Hopefully this gets resolved soon, it got a bit messy.

**Comment by ss37** [comment]
Thanks for the quick reply! Do you have snippet or link for it in the above-mentioned project to reproduce the bug?

**Comment by egilsster** [comment]
I don't, all I did was prefix the `global` variable with an underscore and fixed the references to it. Then in my pre-test npm script I ran `mv` to add `_backup` to the original file in node_modules, then `cp` to copy my patched version  into the node_modules location, and my post-test script renamed the backup file back to overwrite the patched version.

Good luck 🙂

**Comment by bobthekingofegypt** [comment]
I have this same issue when using jest.

@aricart What was the issue with using commonjs_strict?  I just checked out both projects and tried out commonjs_strict in the stan.js gen command

```
"scripts": {
  .....,
  "gen": "protoc --js_out=import_style=commonjs_strict,binary:. lib/pb/protocol.proto",
```

It generates a single file using proto object instead of global.  If I misspell commonjs_strict it will spit out an individual file for each type in the proto file (I originally put in commonjs-strict by mistake), is that what happened? 

It would be great if we could get this issue resolved.

**Comment by aricart** [comment]
@bobthekingofegypt let me try that. Perhaps the new version of the protoc is doing something different?

**Comment by aricart** [comment]
I have released v0.2.6 to npm, containing the fix for this issue.

**Comment by bobthekingofegypt** [comment]
@aricart thanks for the fix.  

Updated and tested the change, my tests appear to pass now.  Will check it properly over the next few days when I back in the office.
